x/stake/keeper/key.go

```diff
@@ -29,169 +28,214 @@ var (
 	binary.BigEndian.PutUint64(heightBytes, ^uint64(validator.BondHeight)) // invert height (older validators first)
 	counterBytes := make([]byte, 2)
 	binary.BigEndian.PutUint16(counterBytes, ^uint16(validator.BondIntraTxCounter)) // invert counter (first txns have priority)
 
-	return append(ValidatorsByPowerIndexKey,
-		append(revokedBytes,
-			append(powerBytes,
-				append(heightBytes, counterBytes...)...)...)...)
+	return append(append(append(append(
+		ValidatorsByPowerIndexKey,
+		revokedBytes...),
+		powerBytes...),
+		heightBytes...),
+		counterBytes...)
 }
 
-// get the key for the accumulated update validators
+// get the key for the accumulated update validators.
+// VALUE: abci.Validator
+// note records using these keys should never persist between blocks
 func GetTendermintUpdatesKey(ownerAddr sdk.Address) []byte {
 	return append(TendermintUpdatesKey, ownerAddr.Bytes()...)
 }
 
 //________________________________________________________________________________
 
-// get the key for delegator bond with validator
-func GetDelegationKey(delegatorAddr, validatorAddr sdk.Address, cdc *wire.Codec) []byte {
-	return append(GetDelegationsKey(delegatorAddr, cdc), validatorAddr.Bytes()...)
+// get the key for delegator bond with validator.
+// VALUE: stake/types.Delegation
+func GetDelegationKey(delegatorAddr, validatorAddr sdk.Address) []byte {
+	return append(GetDelegationsKey(delegatorAddr), validatorAddr.Bytes()...)
 }
 
 // get the prefix for a delegator for all validators
-func GetDelegationsKey(delegatorAddr sdk.Address, cdc *wire.Codec) []byte {
-	res := cdc.MustMarshalBinary(&delegatorAddr)
-	return append(DelegationKey, res...)
+func GetDelegationsKey(delegatorAddr sdk.Address) []byte {
+	return append(DelegationKey, delegatorAddr.Bytes()...)
 }
 
 //________________________________________________________________________________
 
-// get the key for an unbonding delegation
-func GetUBDKey(delegatorAddr, validatorAddr sdk.Address, cdc *wire.Codec) []byte {
-	return append(GetUBDsKey(delegatorAddr, cdc), validatorAddr.Bytes()...)
+// get the key for an unbonding delegation by delegator and validator addr.
+// VALUE: stake/types.UnbondingDelegation
+func GetUBDKey(delegatorAddr, validatorAddr sdk.Address) []byte {
+	return append(
+		GetUBDsKey(delegatorAddr.Bytes()),
+		validatorAddr.Bytes()...)
 }
 
 // get the index-key for an unbonding delegation, stored by validator-index
-func GetUBDByValIndexKey(delegatorAddr, validatorAddr sdk.Address, cdc *wire.Codec) []byte {
-	return append(GetUBDsByValIndexKey(validatorAddr, cdc), delegatorAddr.Bytes()...)
+// VALUE: none (key rearrangement used)
+func GetUBDByValIndexKey(delegatorAddr, validatorAddr sdk.Address) []byte {
+	return append(GetUBDsByValIndexKey(validatorAddr), delegatorAddr.Bytes()...)
+}
+
+// rearrange the ValIndexKey to get the UBDKey
+func GetUBDKeyFromValIndexKey(IndexKey []byte) []byte {
+	addrs := IndexKey[1:] // remove prefix bytes
+	split := len(addrs) / 2
```

Comment:
Presumably we know the exact byte length of each address - should we assert that?

Reply:
agree addressed all